src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/TextDocumentUriPresentationEndpoint.cs

```diff
@@ -0,0 +1,225 @@
+            RazorDocumentMappingService razorDocumentMappingService,
+            RazorComponentSearchEngine razorComponentSearchEngine,
+            ILoggerFactory loggerFactory)
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (razorDocumentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(razorDocumentMappingService));
+            }
+
+            if (razorComponentSearchEngine is null)
+            {
+                throw new ArgumentNullException(nameof(razorComponentSearchEngine));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _razorDocumentMappingService = razorDocumentMappingService;
+            _razorComponentSearchEngine = razorComponentSearchEngine;
+            _logger = loggerFactory.CreateLogger<TextDocumentUriPresentationEndpoint>();
+        }
+
+        public RegistrationExtensionResult GetRegistration()
+        {
+            const string AssociatedServerCapability = "_vs_uriPresentationProvider";
+
+            return new RegistrationExtensionResult(AssociatedServerCapability, true);
+        }
+
+        public async Task<WorkspaceEdit?> Handle(UriPresentationParams request, CancellationToken cancellationToken)
+        {
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            if (request.Range?.Start.TryGetAbsoluteIndex(sourceText, _logger, out var hostDocumentIndex) != true)
+            {
+                return null;
```

Comment:
This would mean that a range had been requested outside of the actual document?

Reply:
I don't even know.. Didn't you write `TryGetAbsoluteIndex` because you were tracking down a bug, and we're going to change them all back to not be Try once the bug is found?

I need to re-write it for VS types soon enough, so I guess I'll find out.